### PR TITLE
Add EnumDisplaySettings

### PR DESCRIPTION
**Pull request by weitzhandler**
This PR adds two functions and their underlying types:

- [`EnumDisplaySettings`](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-enumdisplaysettingsw)
- [`EnumDisplaySettingsEx`](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-enumdisplaysettingsexw)
- [`DEVMODE`](https://docs.microsoft.com/en-us/windows/win32/api/wingdi/ns-wingdi-devmode)

There must be a problem, probably fixable easily, but the test I added [here](https://github.com/dotnet/pinvoke/compare/master...weitzhandler:feature/additional-functions?expand=1#diff-91650ee7838996e1477a028ecc838259) doesn't pass.

@AArnott can you please and have a look and/or checkout my branch?

### CODE REVIEW

File: src/User32/User32.cs
Comment (AArnott) [inline src/User32/User32.cs]: nit: we don't bother with `EntryPoint` in general in this repo.

File: src/Windows.Core/DEVMODEW.cs
Comment (AArnott) [inline src/Windows.Core/DEVMODEW.cs]: I don't think we've ever used the `W` suffix on our structs, any more than we've used it on our p/invoke methods themselves. In native code the headers create `DEVMODE` as an alias, so it would also be accurate to use `DEVMODE` here. Even though it's perhaps less precise, it fits the precedent we have so I would prefer `DEVMODE` for the name.

Similarly, can you rename each enum to remove the `W` from the name?

File: src/Windows.Core/DEVMODEW.cs
Comment (AArnott) [inline src/Windows.Core/DEVMODEW.cs]: This needs to be 64 (and all subsequent fields adjusted accordingly). As it comes after a 32 `char` field, where `char` is 2 bytes.

Tip I learned from @qmfrederik: write a little VC++ program with output such as this:

```c
int main()
{
    std::cout << "dmSpecVersion offset: " << offsetof(DEVMODE, dmSpecVersion) << " size: " << sizeof(DEVMODE::dmSpecVersion) << "\n";
}
```

That prints out:

```
dmSpecVersion offset: 64 size: 2
```

Use this for any field you're not confident of the position or size.
Compile and run your program both for x86 and x64 to compare when in doubt about how pointer size may impact the layout.

File: src/Windows.Core/DEVMODEW.cs
Comment (AArnott) [inline src/Windows.Core/DEVMODEW.cs]: Although the enum values are missing from the docs, I found this in the header files:

```c

/* DEVMODE dmDisplayFlags flags */

// #define DM_GRAYSCALE            0x00000001 /* This flag is no longer valid */
#define DM_INTERLACED           0x00000002
#define DMDISPLAYFLAGS_TEXTMODE 0x00000004
```

They are *flags*, so can we rename this enum to be `DEVMODE_DISPLAYFLAGS` to match the struct name and better fit the member name that uses it?

File: src/Windows.Core/DEVMODEW.cs
Comment (AArnott) [inline src/Windows.Core/DEVMODEW.cs]: Can you add something like this above each enum defined here?

```cs
/// <summary>Defines the values that may be used in the <see cref="DEVMODE.dnNup" /> member.</summary>
```

File: src/Windows.Core/DEVMODEW.cs
Comment (AArnott) [inline src/Windows.Core/DEVMODEW.cs]: Note to self for later (unless you want to try it, @weitzhandler): use regex or multi-caret in VS to mass-convert all these comments to xml doc comments so they are exposed in the API.

File: src/Windows.Core/POINT.cs
Comment (AArnott) [inline src/Windows.Core/POINT.cs]: Please drop this. `Sequential` is C#'s default so we needn't define it.

**Review by AArnott — CHANGES_REQUESTED** [review]
Thanks for your contribution, @weitzhandler! This looks like a lot of work and including the test was a great way to identify the bug in the field offsets.

File: src/User32/User32.cs
Comment (weitzhandler) [reply src/User32/User32.cs]: I thought that in this case it'd be different because we want to explicitly call the wide-char pinvoke.

File: src/Windows.Core/DEVMODEW.cs
Comment (weitzhandler) [reply src/Windows.Core/DEVMODEW.cs]: I thought they they were different in the `DEVMODEW`/`DEVMODE` case but you're right.  
I just rechecked the _wingdi.h_ header and can confirm that. Will rename all `DEVMODEW` instances in my overriding commit.

File: src/User32/User32.cs
Comment (AArnott) [reply src/User32/User32.cs]: We always will anyway. You'd have to go back to Windows 95 and run .NET there for it to default to the Ansi variety, and .NET doesn't even support those OS's any more.

File: src/Windows.Core/DEVMODEW.cs
Comment (weitzhandler) [reply src/Windows.Core/DEVMODEW.cs]: Lovely. I wasn't even aware of the `offsetof` operator in C++!
My C++ sucks anyway, otherwise I'd probably write my apps C++ in the first place...

File: src/Windows.Core/DEVMODEW.cs
Comment (weitzhandler) [reply src/Windows.Core/DEVMODEW.cs]: Yeah. Some of these comments I already converted to XML docs.
I'll also rename accordingly and decorate with `FlagsAttribute`.

BTW, do you want me to place all these enums each into a separate file?

File: src/Windows.Core/DEVMODEW.cs
Comment (weitzhandler) [reply src/Windows.Core/DEVMODEW.cs:366]: Will do.

File: src/Windows.Core/DEVMODEW.cs
Comment (weitzhandler) [reply src/Windows.Core/DEVMODEW.cs]: LOL
I don't mind converting them to proper XML docs so it looks more appealing C#ish.

File: src/User32/User32.cs
Comment (weitzhandler) [reply src/User32/User32.cs]: Wasn't delving into the actual meaning of the version numbers sorry.

File: src/Windows.Core/DEVMODEW.cs
Comment (AArnott) [reply src/Windows.Core/DEVMODEW.cs]: > do you want me to place all these enums each into a separate file?

My first inclination was yes, but given they are very much just to support one struct (and even named to signify that), I think one file is probably best.

File: src/Windows.Core/DEVMODEW.cs
Comment (AArnott) [reply src/Windows.Core/DEVMODEW.cs]: Thanks.
note: It's not just C# syntax that looks better. It's so that someone accessing the enum value from another project will see intellisense doc pop up with the explanation of what it means.

Comment (weitzhandler) [reply src/Windows.Core/DEVMODEW.cs]: Thanks for your prompt replies Andrew!

File: src/Windows.Core/DEVMODEW.cs
Comment (weitzhandler) [reply src/Windows.Core/DEVMODEW.cs]: I've seen some occurrences in PInvoke where the flags are also named in pascal-case. Should this specific one be different? I mean what's the general naming rule for options and flags? How about `DEVMODE_DisplayFlags`?

Comment (AArnott) [reply src/Windows.Core/DEVMODEW.cs]: Our official naming rules are in our CONTRIBUTING.md doc. We add to that as undocumented cases become common. In the interim (like this case, which is the first like it that I've encountered) we muddle through, trying to invent something as consistent with precedent as possible.

We use `Flags` suffixes for flags enums, otherwise Options is common. We always name enums exactly as they are named in the header files, when they are actually defined, but they rarely are. Usually they are just a bunch of loosely bundled `#define` constants.
If we knew that some of these enums were used in other contexts, we'd study those other contexts and promote the enum to a top-level one that doesn't only serve this particular struct.

File: src/Windows.Core/DEVMODEW.cs
Comment (weitzhandler) [reply src/Windows.Core/DEVMODEW.cs]: That's actually the output I got (don't worry, I haven't typed them all manually, I used copy paste, multiline-edit, and find-replace...):

```
dmDeviceName: 0
dmSpecVersion: 32       
dmDriverVersion: 34     
dmSize: 36
dmDriverExtra: 38       
dmFields: 40
dmOrientation: 44       
dmPaperSize: 46
dmPaperLength: 48
dmPaperWidth: 50
dmScale: 52
dmCopies: 54
dmDefaultSource: 56
dmPrintQuality: 58
dmPosition: 44
dmDisplayOrientation: 52
dmDisplayFixedOutput: 56
dmColor: 60
dmDuplex: 62
dmYResolution: 64
dmTTOption: 66
dmCollate: 68
dmFormName: 70
dmLogPixels: 102
dmBitsPerPel: 104
dmPelsWidth: 108
dmPelsHeight: 112
dmDisplayFlags: 116
dmNup: 116
dmDisplayFrequency: 120
dmICMMethod: 124
dmICMIntent: 128
dmMediaType: 132
dmDitherType: 136
dmReserved1: 140
dmReserved2: 144
dmPanningWidth: 148
dmPanningHeight: 152
```

Comment (weitzhandler) [reply src/Windows.Core/DEVMODEW.cs]: Last two offsets don't match with my commit. Will see if that fixes it.

File: src/Windows.Core/DEVMODEW.cs
Comment (weitzhandler) [reply src/Windows.Core/DEVMODEW.cs]: So even this enum, has one value defined as `DM_INTERLACED` while the other is `DMDISPLAYFLAGS_TEXTMODE`...
I wasn't sure what your verdict was, I hope you're fine with _DEVMODE_Display**Flags**_.

File: src/Windows.Core/DEVMODEW.cs
Comment (weitzhandler) [reply src/Windows.Core/DEVMODEW.cs]: It didn't. I'll try checking with 64 for each char array.

Comment (weitzhandler) [reply src/Windows.Core/DEVMODEW.cs]: OK running the `offsetof` with `DEVMODEW` gave me the desired values, and the test passes now.

File: src/Windows.Core/DEVMODEW.cs
Comment (weitzhandler) [reply src/Windows.Core/DEVMODEW.cs]: [Here](https://bit.ly/2WkoDLL)'s how I'm converting the comments into XML docs easily. You may find it useful in the future.
After you load the page, make a change and undo it to load the regex, because it takes slower the first refresh and is blocked.

File: src/Windows.Core/DEVMODE.cs
Comment (vatsan-madhavan) [inline src/Windows.Core/DEVMODE.cs]: Since `DEVMODE[W]` is a `struct`, could structuring this (and other types where this suggestion might apply as well) as a nested-type `DEVMODE.FieldUseFlags` make this a bit cleaner to read/use ? 

I'm usually bit wary of nested public types. They are impossible to type-forward to another assembly. That said, this repo is designed around nested public types, so I figured I'd offer-up this suggestion 😉

File: src/Windows.Core/DEVMODE.cs
Comment (vatsan-madhavan) [inline src/Windows.Core/DEVMODE.cs]: Would it be better for nested types to move into separate files like DEVMODE+TypeName.cs ?

File: src/Windows.Core/DEVMODE.cs
Comment (vatsan-madhavan) [inline src/Windows.Core/DEVMODE.cs]: This enum has as SuppressMessage but others don't....

File: src/Windows.Core/DEVMODE.cs
Comment (vatsan-madhavan) [inline src/Windows.Core/DEVMODE.cs]: With Flags enums, sometimes there is value in having a `None=0x0` value. It depends on how the corresponding Win32 function uses these flags. If the receiving function sometimes uses 0 to indicate no-flags, then introducing a `None` here (and in other Flags enums) could help the consumers of these API's.

**Review by vatsan-madhavan — COMMENTED** [review]
👍

File: src/Windows.Core/DEVMODE.cs
Comment (weitzhandler) [reply src/Windows.Core/DEVMODE.cs]: I think might be too risky, no?
`DEVMODE` is a bit central. It's also used around printers and other devices. Especially given that this is a `struct`, and not a center class like `User32`.
I'm resolving it because we kinda discussed this [here](https://github.com/dotnet/pinvoke/pull/510#discussion_r454363172), feel free to reopen of you feel sentimental about it.

File: src/Windows.Core/DEVMODE.cs
Comment (weitzhandler) [reply src/Windows.Core/DEVMODE.cs]: We discussed this [here](https://github.com/dotnet/pinvoke/pull/510#discussion_r454363172).

File: src/Windows.Core/DEVMODE.cs
Comment (weitzhandler) [reply src/Windows.Core/DEVMODE.cs]: It's only complaining about the first one.

File: src/Windows.Core/DEVMODE.cs
Comment (vatsan-madhavan) [reply src/Windows.Core/DEVMODE.cs]: My question was based on an inference off of the type's name - `DEVMODE_<TypeName>`- which to me suggests that the type is *atleast somewhat* exclusive to `DEVMODE`. If it's indeed exclusive, then nesting would not lead to discovery problems (and it won't affect the layout of the `DEVMODE` struct AFAIK - you can test it). 

OTOH, if these types don't have a strong affinity to `DEVMODE`, then perhaps they don't have to be named this way. They could just be named `<Typename>` and added to the `Windows.Core` `PInvoke` namespace directly.

Comment (weitzhandler) [reply src/Windows.Core/DEVMODE.cs]: It's going to be hard to track down all those types and where they're used in.
There are these two things to consider:
- Skimming through the _wingdi.h_ header (where `DEVMODE` is defined), it doesn't seem there are any other usages for these enums. I don't have the time resources to make proper research to conclude otherwise.
- They are defined globally in the header, so we might consider exposing them as such too.
- Unlike other nested types, these enums are going to be referred to externally, so it will be tedious to call them each time with their full enclosure signature, and adding a `using` alias might not be justified.

File: src/Windows.Core/DEVMODE.cs
Comment (weitzhandler) [reply src/Windows.Core/DEVMODE.cs]: Done.

File: src/Windows.Core/DEVMODE.cs
Comment (vatsan-madhavan) [reply src/Windows.Core/DEVMODE.cs]: Header definitions are all going to be global. Windows headers conventionally everything at a flat global scope. 

If these must be global-ish, then perhaps consider using C# conventions for type names - esp. since these names aren't directly lifted from a Windows header. For e.g., s/`DEVMODE_AbcFlags`/`DevModeAbcFlags`, etc.

Comment (weitzhandler) [reply src/Windows.Core/DEVMODE.cs]: I'm not sure what's the final verdict here.
They're currently all under the `PInvoke` namespace, but they're all called `DEVMODE_AbcFlags`. Do you want me to change them all?

Comment (vatsan-madhavan) [reply src/Windows.Core/DEVMODE.cs]: My comments are suggestions, not requirements 😄 

Where these types go - `PInvoke` namespace vs. nested type inside `DEVMODE` struct is I think a judgment call. I'd defer to your sensibility here since you have researched this plenty. If you think they are better suited under the `PInvoke` namespace then that's where they should go IMHO 😄 

Re: How to name these types: `DEVMODE_EnumName` vs. `DevModeTypeName` - is a matter of style and consistency. There are arguments that support either style. `DEVMODE_EnumName` is an obvious choice since `DEVMODE` is how the Win32 type is styled and using that as a prefix aids in discovery; OTOH `DevModeTypeName` is how C# types are typically styled, and types like [DisplayDeviceFlags](https://github.com/dotnet/pinvoke/blob/master/src/Windows.Core/DISPLAY_DEVICE%2BDisplayDeviceFlags.cs) use this convention (instead of say, naming the enum type [`DISPLAY_DEVICE_StateFlags`](https://github.com/dotnet/pinvoke/blob/86376b1de31a81507b6f6a00aa8d8868fc94665b/src/Windows.Core/DISPLAY_DEVICE.cs#L29))

Comment (weitzhandler) [reply src/Windows.Core/DEVMODE.cs]: Awesome then. The way they're all currently named is `PInvoke.DEVMODE_SomeOptionFlags`. And BTW I wrote the `DISPLAY_DEVICE_StateFlags` 😍, but I did it based on other occurrences I found in PInvoke.
So I guess we're probably good to go, unless I missed something.

Comment (AArnott) [reply src/Windows.Core/DEVMODE.cs]: > I'm usually bit wary of nested public types. They are impossible to type-forward to another assembly. That said, this repo is designed around nested public types, so I figured I'd offer-up this suggestion 😉

See https://github.com/dotnet/pinvoke/issues/288#issuecomment-652750274. Although we've traditionally put all structs as nested under the library class, I think we should reverse that design and put *most* structs directly under `PInvoke`, including most existing ones in a breaking API change when we're ready to make such a change. For now, new structs can follow the new policy, IMO. 
However, where a struct or enum is so coupled to another struct such that you'd name one after the other, I think nesting makes sense. It just changes `DEVMODE_SomeOptionFlags` to `DEVMODE.SomeOptionFlags` and in the meantime removes `*SomeOptionFlags` from the completion lists of other scenarios, which seems like goodness, IMO. 

That leaves the question of whether it should simply be `SomeOptionsFlags` vs. `DEVMODE*SomeOptionFlags`. For that, as has also been expressed here, if we knew that `SomeOptionFlags` was actually used by other APIs, I'd favor making them more prominently named/placed. But if we're reasonably confident that it only applies to one struct, I'd keep it named as if it only applied there.
If the worst happened (i.e. we discovered another API that uses it too), we can always simply define another enum with the same values but with a more prominent name.

File: src/User32/User32+EnumDisplaySettingsExFlags.cs
Comment (AArnott) [inline src/User32/User32+EnumDisplaySettingsExFlags.cs]: nit: when specifying the overload, prefer the original one (the one with pointers).
Not a big deal nor something you have to correct on this PR, but something for next time.

File: src/User32/User32.cs
Comment (AArnott) [inline src/User32/User32.cs]: We usually discourage `string`, preferring instead to match the header file, in this case:

```cs
[Friendly(FriendlyFlags.Array | FriendlyFlags.In)] char* lpszDeviceName,
```

At the moment, that's a little less convenient when you have a `string` you want to pass in, but #444 and #286 will take care of this.
And in particular since the input to this probably comes from `DISPLAY_DEVICE.DeviceName` which is a `char*` already, accepting it here may mean you never have to allocate a `string` in the first place.
Same for other p/invoke methods added here.

If you look, I'm sure you'll find some examples of using `string` elsewhere in this repo, including from my own contributions. Most of these were from early in the repo's development before we'd standardized on matching the header files as closely as possible.

File: src/Windows.Core/DEVMODE.cs
Comment (AArnott) [reply src/Windows.Core/DEVMODE.cs]: Looking over this in general, this leads me to believe the only change to make here is to move these `DEVMODE_` types into `DEVMODE` as nested types and trim the prefix from them, since their existing names already commit us to consider these tightly coupled.

**Review by AArnott — CHANGES_REQUESTED** [review]
It looks good. We could take it as-is. But if you don't mind humoring me a bit as we negotiate the patterns for nested types (either in discussion or applying my suggested changes), I'd like to explore that option first.

**Comment by weitzhandler** [comment]
> I'd like to explore that option first

So I added it as a separate commit, so we can roll back if you don't like it.

**Review by AArnott — APPROVED** [review]
Thanks!

**Pull request merged**
